File: user-product-service/src/controllers/productController.js
const { Op } = require('sequelize');
const { Product, Category } = require('../models');

// Helper to create slug
const slugify = (text) => {
  return text
    .toString()
    .toLowerCase()
    .trim()
    .replace(/\s+/g, '-')
    .replace(/[^\w\-]+/g, '')
    .replace(/\-\-+/g, '-');
};

// @desc    Fetch all products with search, filter, sort, and pagination
// @route   GET /api/products
// @access  Public
const getProducts = async (req, res) => {
  try {
    const page = parseInt(req.query.page, 10) || 1;
    const limit = parseInt(req.query.limit, 10) || 12;
    const offset = (page - 1) * limit;

    const { search, q, category, minPrice, maxPrice, sort, featured, inStock } = req.query;

    const whereClause = {};

    // Search filter
    const keyword = search || q;
    if (keyword) {
      whereClause[Op.or] = [
        { name: { [Op.like]: `%${keyword}%` } },
        { description: { [Op.like]: `%${keyword}%` } },
        { category: { [Op.like]: `%${keyword}%` } },
      ];
    }

    // Category filter
    if (category && category !== 'All' && category !== 'all') {
      whereClause.category = category;
    }

    // Price range
    if (minPrice || maxPrice) {
      whereClause.price = {};
      if (minPrice) whereClause.price[Op.gte] = parseFloat(minPrice);
      if (maxPrice) whereClause.price[Op.lte] = parseFloat(maxPrice);
    }

    // Featured flag
    if (featured === 'true' || featured === true) {
      whereClause.is_featured = true;
    }

    // In-Stock filter
    if (inStock === 'true' || inStock === true) {
      whereClause.stock = { [Op.gt]: 0 };
    }

    // Sorting
    let orderClause = [['created_at', 'DESC']];
    if (sort === 'price_asc') {
      orderClause = [['price', 'ASC']];
    } else if (sort === 'price_desc') {
      orderClause = [['price', 'DESC']];
    } else if (sort === 'rating') {
      orderClause = [['rating', 'DESC']];
    } else if (sort === 'name_asc') {
      orderClause = [['name', 'ASC']];
    } else if (sort === 'oldest') {
      orderClause = [['created_at', 'ASC']];
    }

    const { count, rows: products } = await Product.findAndCountAll({
      where: whereClause,
      order: orderClause,
      limit,
      offset,
      include: [
        {
          model: Category,
          as: 'categoryDetails',
          attributes: ['id', 'name', 'slug'],
        },
      ],
    });

    res.json({
      success: true,
      total: count,
      page,
      pages: Math.ceil(count / limit),
      limit,
      products,
    });
  } catch (error) {
    console.error('Error in getProducts:', error);
    res.status(500).json({
      success: false,
      message: error.message || 'Error fetching products',
    });
  }
};

// @desc    Fetch single product by ID
// @route   GET /api/products/:id
// @access  Public
const getProductById = async (req, res) => {
  try {
    const product = await Product.findByPk(req.params.id, {
      include: [
        {
          model: Category,
          as: 'categoryDetails',
          attributes: ['id', 'name', 'slug', 'description'],
        },
      ],
    });

    if (!product) {
      return res.status(404).json({
        success: false,
        message: `Product with ID ${req.params.id} not found`,
      });
    }

    res.json({
      success: true,
      product,
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message || 'Error fetching product details',
    });
  }
};

// @desc    Create a new product
// @route   POST /api/products
// @access  Private/Admin
const createProduct = async (req, res) => {
  try {
    const {
      name,
      description,
      price,
      discount_price,
      image,
      category,
      category_id,
      stock,
      rating,
      is_featured,
    } = req.body;

    if (!name || !description || price === undefined || !image || !category) {
      return res.status(400).json({
        success: false,
        message: 'Please provide all required product fields: name, description, price, image, category',
      });
    }

    const slug = slugify(name) + '-' + Date.now().toString().slice(-4);

    const product = await Product.create({
      name,
      slug,
      description,
      price: parseFloat(price),
      discount_price: discount_price ? parseFloat(discount_price) : null,
      image,
      category,
      category_id: category_id || null,
      stock: parseInt(stock, 10) || 0,
      rating: rating ? parseFloat(rating) : 4.5,
      is_featured: is_featured || false,
    });

    res.status(201).json({
      success: true,
      message: 'Product created successfully',
      product,
    });
  } catch (error) {
    console.error('Error creating product:', error);
    res.status(500).json({
      success: false,
      message: error.message || 'Error creating product',
    });
  }
};

// @desc    Update an existing product
// @route   PUT /api/products/:id
// @access  Private/Admin
const updateProduct = async (req, res) => {
  try {
    const product = await Product.findByPk(req.params.id);

    if (!product) {
      return res.status(404).json({
        success: false,
        message: 'Product not found',
      });
    }

    const {
      name,
      description,
      price,
      discount_price,
      image,
      category,
      category_id,
      stock,
      rating,
      is_featured,
    } = req.body;

    if (name) {
      product.name = name;
      product.slug = slugify(name);
    }
    if (description !== undefined) product.description = description;
    if (price !== undefined) product.price = parseFloat(price);
    if (discount_price !== undefined) product.discount_price = discount_price ? parseFloat(discount_price) : null;
    if (image !== undefined) product.image = image;
    if (category !== undefined) product.category = category;
    if (category_id !== undefined) product.category_id = category_id;
    if (stock !== undefined) product.stock = parseInt(stock, 10);
    if (rating !== undefined) product.rating = parseFloat(rating);
    if (is_featured !== undefined) product.is_featured = is_featured;

    await product.save();

    res.json({
      success: true,
      message: 'Product updated successfully',
      product,
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message || 'Error updating product',
    });
  }
};

// @desc    Delete a product
// @route   DELETE /api/products/:id
// @access  Private/Admin
const deleteProduct = async (req, res) => {
  try {
    const product = await Product.findByPk(req.params.id);

    if (!product) {
      return res.status(404).json({
        success: false,
        message: 'Product not found',
      });
    }

    await product.destroy();

    res.json({
      success: true,
      message: 'Product deleted successfully',
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message || 'Error deleting product',
    });
  }
};

// @desc    Reduce product stock (for Order service cross-call)
// @route   POST /api/products/reduce-stock
// @access  Public / Internal Service
const reduceStock = async (req, res) => {
  try {
    const { items } = req.body; // array of { product_id, quantity }

    if (!items || !Array.isArray(items)) {
      return res.status(400).json({
        success: false,
        message: 'Items array is required',
      });
    }

    for (const item of items) {
      const product = await Product.findByPk(item.product_id);
      if (product) {
        product.stock = Math.max(0, product.stock - (parseInt(item.quantity, 10) || 1));
        await product.save();
      }
    }

    res.json({
      success: true,
      message: 'Stock updated successfully',
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message || 'Error reducing product stock',
    });
  }
};

module.exports = {
  getProducts,
  getProductById,
  createProduct,
  updateProduct,
  deleteProduct,
  reduceStock,
};

File: user-product-service/src/main/java/com/nexura/userproduct/controller/ProductController.java
package com.nexura.userproduct.controller;

import com.nexura.userproduct.dto.ProductRequest;
import com.nexura.userproduct.dto.StockReduceRequest;
import com.nexura.userproduct.model.Product;
import com.nexura.userproduct.repository.ProductRepository;
import jakarta.persistence.criteria.Predicate;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/products")
public class ProductController {

    @Autowired
    private ProductRepository productRepository;

    private String slugify(String text) {
        if (text == null) return "";
        return text.toLowerCase().trim().replaceAll("\\s+", "-").replaceAll("[^\\w-]+", "");
    }

    @GetMapping
    public ResponseEntity<?> getProducts(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "12") int limit,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String q,
            @RequestParam(required = false) String category,
            @RequestParam(required = false) Double minPrice,
            @RequestParam(required = false) Double maxPrice,
            @RequestParam(required = false) String sort,
            @RequestParam(required = false) Boolean featured,
            @RequestParam(required = false) Boolean inStock) {

        int pageNumber = Math.max(0, page - 1);
        int pageSize = Math.max(1, limit);

        Sort sortOrder = Sort.by(Sort.Direction.DESC, "createdAt");
        if ("price_asc".equalsIgnoreCase(sort)) {
            sortOrder = Sort.by(Sort.Direction.ASC, "price");
        } else if ("price_desc".equalsIgnoreCase(sort)) {
            sortOrder = Sort.by(Sort.Direction.DESC, "price");
        } else if ("rating".equalsIgnoreCase(sort)) {
            sortOrder = Sort.by(Sort.Direction.DESC, "rating");
        } else if ("name_asc".equalsIgnoreCase(sort)) {
            sortOrder = Sort.by(Sort.Direction.ASC, "name");
        } else if ("oldest".equalsIgnoreCase(sort)) {
            sortOrder = Sort.by(Sort.Direction.ASC, "createdAt");
        }

        Pageable pageable = PageRequest.of(pageNumber, pageSize, sortOrder);

        String keyword = search != null ? search : q;

        Specification<Product> spec = (root, query, criteriaBuilder) -> {
            List<Predicate> predicates = new ArrayList<>();

            if (keyword != null && !keyword.trim().isEmpty()) {
                String pattern = "%" + keyword.trim().toLowerCase() + "%";
                Predicate nameLike = criteriaBuilder.like(criteriaBuilder.lower(root.get("name")), pattern);
                Predicate descLike = criteriaBuilder.like(criteriaBuilder.lower(root.get("description")), pattern);
                Predicate catLike = criteriaBuilder.like(criteriaBuilder.lower(root.get("category")), pattern);
                predicates.add(criteriaBuilder.or(nameLike, descLike, catLike));
            }

            if (category != null && !category.trim().isEmpty() && !"all".equalsIgnoreCase(category)) {
                predicates.add(criteriaBuilder.equal(criteriaBuilder.lower(root.get("category")), category.trim().toLowerCase()));
            }

            if (minPrice != null) {
                predicates.add(criteriaBuilder.greaterThanOrEqualTo(root.get("price"), minPrice));
            }

            if (maxPrice != null) {
                predicates.add(criteriaBuilder.lessThanOrEqualTo(root.get("price"), maxPrice));
            }

            if (Boolean.TRUE.equals(featured)) {
                predicates.add(criteriaBuilder.equal(root.get("isFeatured"), true));
            }

            if (Boolean.TRUE.equals(inStock)) {
                predicates.add(criteriaBuilder.greaterThan(root.get("stock"), 0));
            }

            return criteriaBuilder.and(predicates.toArray(new Predicate[0]));
        };

        Page<Product> productPage = productRepository.findAll(spec, pageable);

        Map<String, Object> response = new HashMap<>();
        response.put("success", true);
        response.put("total", productPage.getTotalElements());
        response.put("page", page);
        response.put("pages", productPage.getTotalPages());
        response.put("limit", pageSize);
        response.put("products", productPage.getContent());

        return ResponseEntity.ok(response);
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getProductById(@PathVariable Long id) {
        Optional<Product> productOpt = productRepository.findById(id);
        if (productOpt.isEmpty()) {
            Map<String, Object> error = new HashMap<>();
            error.put("success", false);
            error.put("message", "Product with ID " + id + " not found");
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
        }

        Map<String, Object> response = new HashMap<>();
        response.put("success", true);
        response.put("product", productOpt.get());

        return ResponseEntity.ok(response);
    }

    @PostMapping
    public ResponseEntity<?> createProduct(@Valid @RequestBody ProductRequest request) {
        String slug = slugify(request.getName()) + "-" + (System.currentTimeMillis() % 10000);

        Product product = new Product();
        product.setName(request.getName());
        product.setSlug(slug);
        product.setDescription(request.getDescription());
        product.setPrice(request.getPrice());
        product.setDiscountPrice(request.getDiscountPrice());
        product.setImage(request.getImage());
        product.setCategory(request.getCategory());
        product.setCategoryId(request.getCategoryId());
        product.setStock(request.getStock() != null ? request.getStock() : 0);
        product.setRating(request.getRating() != null ? request.getRating() : 4.5);
        product.setNumReviews(0);
        product.setIsFeatured(request.getIsFeatured() != null ? request.getIsFeatured() : false);

        Product savedProduct = productRepository.save(product);

        Map<String, Object> response = new HashMap<>();
        response.put("success", true);
        response.put("message", "Product created successfully");
        response.put("product", savedProduct);

        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateProduct(@PathVariable Long id, @RequestBody ProductRequest request) {
        Optional<Product> productOpt = productRepository.findById(id);
        if (productOpt.isEmpty()) {
            Map<String, Object> error = new HashMap<>();
            error.put("success", false);
            error.put("message", "Product not found");
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
        }

        Product product = productOpt.get();

        if (request.getName() != null && !request.getName().isBlank()) {
            product.setName(request.getName());
            product.setSlug(slugify(request.getName()));
        }
        if (request.getDescription() != null) {
            product.setDescription(request.getDescription());
        }
        if (request.getPrice() != null) {
            product.setPrice(request.getPrice());
        }
        if (request.getDiscountPrice() != null) {
            product.setDiscountPrice(request.getDiscountPrice());
        }
        if (request.getImage() != null) {
            product.setImage(request.getImage());
        }
        if (request.getCategory() != null) {
            product.setCategory(request.getCategory());
        }
        if (request.getCategoryId() != null) {
            product.setCategoryId(request.getCategoryId());
        }
        if (request.getStock() != null) {
            product.setStock(request.getStock());
        }
        if (request.getRating() != null) {
            product.setRating(request.getRating());
        }
        if (request.getIsFeatured() != null) {
            product.setIsFeatured(request.getIsFeatured());
        }

        Product savedProduct = productRepository.save(product);

        Map<String, Object> response = new HashMap<>();
        response.put("success", true);
        response.put("message", "Product updated successfully");
        response.put("product", savedProduct);

        return ResponseEntity.ok(response);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteProduct(@PathVariable Long id) {
        Optional<Product> productOpt = productRepository.findById(id);
        if (productOpt.isEmpty()) {
            Map<String, Object> error = new HashMap<>();
            error.put("success", false);
            error.put("message", "Product not found");
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
        }

        productRepository.delete(productOpt.get());

        Map<String, Object> response = new HashMap<>();
        response.put("success", true);
        response.put("message", "Product deleted successfully");

        return ResponseEntity.ok(response);
    }

    @PostMapping("/reduce-stock")
    public ResponseEntity<?> reduceStock(@RequestBody StockReduceRequest request) {
        if (request == null || request.getItems() == null) {
            Map<String, Object> error = new HashMap<>();
            error.put("success", false);
            error.put("message", "Items array is required");
            return ResponseEntity.badRequest().body(error);
        }

        for (StockReduceRequest.StockItem item : request.getItems()) {
            if (item.getProductId() != null) {
                Optional<Product> productOpt = productRepository.findById(item.getProductId());
                if (productOpt.isPresent()) {
                    Product product = productOpt.get();
                    int qty = item.getQuantity() != null ? item.getQuantity() : 1;
                    product.setStock(Math.max(0, product.getStock() - qty));
                    productRepository.save(product);
                }
            }
        }

        Map<String, Object> response = new HashMap<>();
        response.put("success", true);
        response.put("message", "Stock updated successfully");

        return ResponseEntity.ok(response);
    }

    @PostMapping("/restock")
    public ResponseEntity<?> restockProduct(@RequestBody StockReduceRequest request) {
        if (request == null || request.getItems() == null) {
            Map<String, Object> error = new HashMap<>();
            error.put("success", false);
            error.put("message", "Items array is required");
            return ResponseEntity.badRequest().body(error);
        }

        for (StockReduceRequest.StockItem item : request.getItems()) {
            if (item.getProductId() != null) {
                Optional<Product> productOpt = productRepository.findById(item.getProductId());
                if (productOpt.isPresent()) {
                    Product product = productOpt.get();
                    int qty = item.getQuantity() != null ? item.getQuantity() : 1;
                    product.setStock(product.getStock() + qty);
                    productRepository.save(product);
                }
            }
        }

        Map<String, Object> response = new HashMap<>();
        response.put("success", true);
        response.put("message", "Stock restored successfully");

        return ResponseEntity.ok(response);
    }
}
